### ISSUE TITLE
System.Text.Json is excruciatingly slow

### ISSUE THREAD

**Issue by zachsaw** [dotnet/runtime#31493]
Tested it on macOS High Sierra with .net core 3.0.100.

Even running it with dotnet run -c Release (or without for Debug mode) with terminal shows that it runs **6x slower than Newtonsoft.Json**!

I've created a repo to demonstrate the problem here - https://github.com/zachsawcs/SerializerBattle

**Comment by ahsonkhan** [comment]
> I'll try to compile a standalone later when I get some time.

@zachsawcs / @zachsaw did you get a chance to create and share the standalone app that shows the performance issue?

**Comment by zachsaw** [comment]
https://www.dropbox.com/s/n7fsxns58a2zqz2/SerializerBattle.zip?dl=0

**Comment by layomia** [comment]
We've made various performance improvements in 5.0 - https://github.com/dotnet/runtime/pulls?q=is%3Apr+label%3Atenet-performance+label%3Aarea-System.Text.Json+is%3Aclosed. Please feel to reopen if you encounter a reproducible performance issue.

**Comment by zachsaw** [comment]
Just to confirm this is now fixed on 5.0. Tested with preview 8. System.Text.Json is still the slower than Newtonsoft, but at least it is no longer > 5x slower.

Just out of curiosity - what was the problem with .net 3.0 and .net 3.1?

```
NetCore.Json Test
=================
Warm-up
Running
Result: 29.373 ms/op

Newtonsoft.Json Test
====================
Warm-up
Running
Result: 28.373 ms/op

Binaron.Serializer Test
======================
Warm-up
Running
Result: 6.827 ms/op
```

**Comment by zachsaw** [comment]
Interestingly, 5.0 seems to have improved the performance of Binaron.Serializer by 15% in this particular test whereas Newtonsoft.Json remained the same. Loving .net 5.0 already!

**Comment by steveharter** [comment]
> Just to confirm this is now fixed on 5.0. Tested with preview 8. System.Text.Json is still the slower than Newtonsoft, but at least it is no longer > 5x slower.

@zachsaw I decompiled your tests and the likely reason for the slowness is due to using polymorphism via `Dictionary<string, object>` for the "Book.Metadata" property. Unlike Newtonsoft, STJ doesn't "guess" at what the `object` type maps to and instead creates a `JsonElement` to represent. This is slow because there is an additional alloc per `object` instance as well as other overhead of `JsonElement`. This was [made faster in 3.1](https://github.com/dotnet/runtime/issues/31493) and any future optimizations are currently this is being tracked in https://github.com/dotnet/runtime/issues/38637.

The "work-around" if you want to call it that is to use a strongly-typed dictionary such as `Dictionary<string, string>` or create a [custom converter that emulates Newtonsoft's behavior](https://github.com/dotnet/runtime/issues/29960#issuecomment-535166692). In a simple local test, adding a single polymorphic dictionary similar to you have it makes performance almost ~2x as slow in STJ.
